File: 2017/ljd/ljd.py
import numpy as np
import matplotlib.pyplot as plt

def cellpotential(r, sigma, epsilon, z, R, a=0.0):
    """
    Lennard-Jones--Devonshire cellpotential function.

    z: coordination number == number of water molecules on the cage surface.
    R: free cavity radius; 3.908, 4.326 AA for 12- and 14-hedral cages, respectively.
    r: position of the guest from the center.
    """
    def delta(N):
        return ((1 - r / R - a / R)**(-N) - (1 + r / R - a / R)**(-N)) / N

    ans = 2 * z * epsilon * (sigma**12 / (R**11 * r) * (delta(10) + a * delta(
        11) / R) - sigma**6 / (R**5 * r) * (delta(4) + a * delta(5) / R))
    return ans

def fvalue(R, sigma, epsilon, beta, z=20):
    r = np.linspace(0,R,100)[1:99]
    plt.plot(r, cellpotential(r, sigma, epsilon, z, R)*beta)
    # plt.plot(r, np.exp(-cellpotential(r, sigma, epsilon, z, R) * beta)*r**2)
    # print(z, np.trapz(np.exp(-beta * cellpotential(r, sigma, epsilon, z, R)) * r**2, dx=r[1]-r[0]))
    return -np.log(4 * np.pi * np.trapz(np.exp(-beta * cellpotential(r, sigma, epsilon, z, R)) * r**2 * 1e-30, dx=r[1]-r[0])) / beta


# 2022-03-25とりあえず書いただけ。まだ使っていない。
# 昔、これとは全く別に検証したはずだが、どこにあるかわからない。
# LJDを使ったとしても、数値積分は必要なので、実際効率がいいのかどうかはわからん。
# でも、単原子分子ならリアルタイムでfを計算できるようになる見込み。
# Devonshireの計算は、ケージより外のホスト格子のことは考慮していないのに、けっこうきれいに
# 全原子計算と合うということは、全原子計算もケージまででいいのかもしれないね。

File: svelte-app/src/lib/ljd/ljd.ts
/**
 * Lennard-Jones--Devonshire cellpotential function.
 * @param r position of the guest from the center (Angstrom)
 * @param sigma sigma (Angstrom)
 * @param epsilon epsilon (kJ/mol)
 * @param z coordination number
 * @param R free cavity radius (Angstrom)
 * @param a shift (default 0.0)
 */
export function cellpotential(
  r: number,
  sigma: number,
  epsilon: number,
  z: number,
  R: number,
  a: number = 0.0
): number {
  const delta = (N: number): number => {
    return (Math.pow(1 - r / R - a / R, -N) - Math.pow(1 + r / R - a / R, -N)) / N;
  };

  const ans =
    2 *
    z *
    epsilon *
    (Math.pow(sigma, 12) / (Math.pow(R, 11) * r) * (delta(10) + (a * delta(11)) / R) -
      Math.pow(sigma, 6) / (Math.pow(R, 5) * r) * (delta(4) + (a * delta(5)) / R));
  return ans;
}

/**
 * Calculate the free energy of cage occupation using LJD approximation.
 * @param R radius
 * @param z number of water molecules
 * @param sigma sigma
 * @param epsilon epsilon
 * @param beta beta (1/NkB*T)
 */
export function fvalue2(
  R: number,
  z: number,
  sigma: number,
  epsilon: number,
  beta: number
): number {
  const n = 100;
  const dr = R / (n - 1);
  const r = Array.from({ length: n }, (_, i) => i * dr);
  const integrand = new Float64Array(n);

  for (let i = 0; i < n; i++) {
    if (i === 0) {
      // Limit as r -> 0 for cellpotential
      // The cellpotential has a 1/r factor, but delta(N) also has r.
      // For r=0, cellpotential(0) = 2*z*epsilon * [ sigma^12/R^12 - sigma^6/R^6 ]
      // Wait, let's check the limit.
      // (1-x)^-N - (1+x)^-N = [1 + Nx + N(N+1)/2 x^2 + ...] - [1 - Nx + N(N+1)/2 x^2 - ...] = 2Nx + O(x^3)
      // So delta(N) / r = 2/R + O(r^2)
      // cellpotential(0) = 2*z*epsilon * [ sigma^12/R^12 - sigma^6/R^6 ]
      const cp0 = 2 * z * epsilon * (Math.pow(sigma / R, 12) - Math.pow(sigma / R, 6));
      integrand[i] = Math.exp(-beta * cp0) * Math.pow(r[i], 2) * 1e-30;
    } else if (i < n - 1) {
      const cp = cellpotential(r[i], sigma, epsilon, z, R);
      integrand[i] = Math.exp(-beta * cp) * Math.pow(r[i], 2) * 1e-30;
    } else {
      // r = R, potential is infinite
      integrand[i] = 0;
    }
  }

  // Trapezoidal rule
  let integral = 0;
  for (let i = 0; i < n - 1; i++) {
    integral += (integrand[i] + integrand[i + 1]) * dr / 2;
  }

  return -Math.log(4 * Math.PI * integral) / beta;
}
